### ISSUE TITLE
Multi-class Random forest support

### ISSUE THREAD

**Issue by StrikerRUS** [Microsoft/LightGBM#881]
I suppose I missed to specify some parameter, but I can't figure which one.

```
from lightgbm.sklearn import LGBMClassifier
from sklearn.datasets import load_iris

iris = load_iris()
X = iris.data
y = iris.target

clf = LGBMClassifier(boosting_type='rf', bagging_fraction=0.5, feature_fraction=0.5)
clf.fit(X, y)
```

Error:
```
---------------------------------------------------------------------------
LightGBMError                             Traceback (most recent call last)
<ipython-input-1-734db61b74f4> in <module>()
      7 
      8 clf = LGBMClassifier(boosting_type='rf', bagging_fraction=0.5, feature_fraction=0.5)
----> 9 clf.fit(X, y)

C:\Program Files\Anaconda3\lib\site-packages\lightgbm\sklearn.py in fit(self, X, y, sample_weight, init_score, eval_set, eval_names, eval_sample_weight, eval_init_score, eval_metric, early_stopping_rounds, verbose, feature_name, categorical_feature, callbacks)
    557                                         verbose=verbose, feature_name=feature_name,
    558                                         categorical_feature=categorical_feature,
--> 559                                         callbacks=callbacks)
    560         return self
    561 

C:\Program Files\Anaconda3\lib\site-packages\lightgbm\sklearn.py in fit(self, X, y, sample_weight, init_score, group, eval_set, eval_names, eval_sample_weight, eval_init_score, eval_group, eval_metric, early_stopping_rounds, verbose, feature_name, categorical_feature, callbacks)
    405                               verbose_eval=verbose, feature_name=feature_name,
    406                               categorical_feature=categorical_feature,
--> 407                               callbacks=callbacks)
    408 
    409         if evals_result:

C:\Program Files\Anaconda3\lib\site-packages\lightgbm\engine.py in train(params, train_set, num_boost_round, valid_sets, valid_names, fobj, feval, init_model, feature_name, categorical_feature, early_stopping_rounds, evals_result, verbose_eval, learning_rates, keep_training_booster, callbacks)
    178     """construct booster"""
    179     try:
--> 180         booster = Booster(params=params, train_set=train_set)
    181         if is_valid_contain_train:
    182             booster.set_train_data_name(train_data_name)

C:\Program Files\Anaconda3\lib\site-packages\lightgbm\basic.py in __init__(self, params, train_set, model_file, silent)
   1251                 train_set.construct().handle,
   1252                 c_str(params_str),
-> 1253                 ctypes.byref(self.handle)))
   1254             """save reference to data"""
   1255             self.train_set = train_set

C:\Program Files\Anaconda3\lib\site-packages\lightgbm\basic.py in _safe_call(ret)
     46     """
     47     if ret != 0:
---> 48         raise LightGBMError(_LIB.LGBM_GetLastError())
     49 
     50 

LightGBMError: b'Check failed: num_tree_per_iteration_ == 1 at d:\\users\\nekit\\downloads\\test\\lightgbm\\python-package\\lightgbm\\src\\boosting\\rf.hpp, line 41 .\n'
```

All params:
```
{'bagging_fraction': 0.5,
 'boosting_type': 'rf',
 'colsample_bytree': 1,
 'feature_fraction': 0.5,
 'learning_rate': 0.1,
 'max_bin': 255,
 'max_depth': -1,
 'min_child_samples': 10,
 'min_child_weight': 5,
 'min_split_gain': 0,
 'n_estimators': 10,
 'n_jobs': -1,
 'num_leaves': 31,
 'objective': 'multiclass',
 'random_state': 0,
 'reg_alpha': 0,
 'reg_lambda': 0,
 'silent': True,
 'subsample': 1,
 'subsample_for_bin': 50000,
 'subsample_freq': 1}
```

**Comment by guolinke** [comment]
@StrikerRUS It doesn't support multi-class RF now.

**Comment by StrikerRUS** [comment]
@guolinke 
Many thanks for your fast feedback!

**Comment by sbodenstein** [comment]
@guolinke: is there any plan on adding this? And what sort of timeline is expected?

**Comment by econkc** [comment]
Hi,

I have the same error message even using gbdt or dart boosting. At first , my data has 6 target classes and it ran fine. However, I have decided to drop one of the classes and now have 5 targets, I change 'num_class' param accordingly but now I have got this error

> LightGBMError: b'Check failed: index < num_tree_per_iteration_ at /Users/travis/miniconda3/conda-bld/lightgbm_1508563794857/work/lightgbm-2.0.10/compile/src/boosting/gbdt.cpp, line 131 .\n'

Any suggestion?

**Comment by guolinke** [comment]
fixed in #1637
